cmd/snap/inhibit.go

```diff
@@ -34,143 +34,153 @@ import (
-	// gateAutoRefreshHookHandler in hooks.go).
-	// waitInhibitUnlock will return also on HintNotInhibited.
-	notInhibited, err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh)
+	hintFlock, err := runinhibitWaitWhileInhibited(snapName, nil, inhibited, 500*time.Millisecond)
 	if err != nil {
+		// It is fine to return an error here without finishing the notification
+		// flow because we either failed because of it or before it, so it
+		// should not have started in the first place.
 		return err
 	}
-	if notInhibited {
-		return nil
-	}
 
-	if isGraphicalSession() {
-		notifiedDesktopIntegration, err := tryNotifyRefreshViaSnapDesktopIntegrationFlow(snapName)
-		if err != nil {
+	// XXX: closing as we don't need it for now, this lock will be used in a later iteration
+	hintFlock.Close()
+
+	if notified {
+		if err := flow.Finish(); err != nil {
 			return err
 		}
-		if notifiedDesktopIntegration {
-			return nil
-		}
-		return graphicalSessionFlow(snapName, hint)
 	}
-	// terminal and headless
-	return textFlow(snapName, hint)
+
+	return nil
 }
 
-func inhibitMessage(snapName string, hint runinhibit.Hint) string {
-	switch hint {
-	case runinhibit.HintInhibitedForRefresh:
-		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
-	default:
-		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
-	}
+type inhibitionFlow interface {
+	Notify() error
+	Finish() error
 }
 
-var isGraphicalSession = func() bool {
-	// TODO: uncomment once there is a proper UX review
-	//return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
-	return false
+type textFlow struct {
+	snapName string
+
+	pb progress.Meter
 }
 
-var pendingRefreshNotification = func(refreshInfo *client.PendingSnapRefreshInfo) error {
-	userclient := client.NewForUids(os.Getuid())
-	if err := userclient.PendingRefreshNotification(context.TODO(), refreshInfo); err != nil {
+func (tf *textFlow) Notify() error {
+	_, err := fmt.Fprintf(Stdout, i18n.G("snap package %q is being refreshed, please wait\n"), tf.snapName)
+	if err != nil {
 		return err
 	}
+	tf.pb = progress.MakeProgressBar(Stdout)
+	tf.pb.Spin(i18n.G("please wait..."))
 	return nil
 }
 
-var finishRefreshNotification = func(refreshInfo *client.FinishedSnapRefreshInfo) error {
-	userclient := client.NewForUids(os.Getuid())
-	if err := userclient.FinishRefreshNotification(context.TODO(), refreshInfo); err != nil {
-		return err
+func (tf *textFlow) Finish() error {
+	if tf.pb != nil {
+		tf.pb.Finished()
 	}
 	return nil
 }
 
-var tryNotifyRefreshViaSnapDesktopIntegrationFlow = func(snapName string) (bool, error) {
+type graphicalFlow struct {
+	snapName string
+
+	notifiedDesktopIntegration bool
+}
+
+func (gf *graphicalFlow) Notify() error {
+	gf.notifiedDesktopIntegration = tryNotifyRefreshViaSnapDesktopIntegrationFlow(gf.snapName)
+	if gf.notifiedDesktopIntegration {
+		return nil
```

Comment:
this case is not tested, maybe wasn't either on master

Reply:
thanks for catching this, updated